pydantic_ai_slim/pydantic_ai/toolsets/_dynamic.py

```diff
@@ -17,23 +16,52 @@
 """A sync/async function which takes a run context and returns a toolset."""
 
 
-@dataclass
 class DynamicToolset(AbstractToolset[AgentDepsT]):
     """A toolset that dynamically builds a toolset using a function that takes the run context.
 
     It should only be used during a single agent run as it stores the generated toolset.
-    To use it multiple times, copy it using `dataclasses.replace`.
+    To use it multiple times, use the `copy()` method.
     """
 
     toolset_func: ToolsetFunc[AgentDepsT]
-    per_run_step: bool = True
-
-    _toolset: AbstractToolset[AgentDepsT] | None = None
-    _run_step: int | None = None
+    per_run_step: bool
+    _id: str | None
+    _toolset: AbstractToolset[AgentDepsT] | None
+    _run_step: int | None
```

Comment:
I wonder if we could solve this "reuse a dynamic toolset across multiple agent runs" issue not by copying it for each run (which is tricky as we see here), but by storing the `_toolset` and `_run_step` on a dict keyed by `ctx.run_id`.

That way, we can keep using the same toolset instance, but state from different runs wouldn't interfere. I think that's worth trying in this PR, as an alternative to the new `copy` method